libbeat/cmd/instance/locks/lock.go

```diff
@@ -0,0 +1,215 @@
+		}
+	}
+
+	// This will be a shared/read lock,
+	// if we need to manage a previous lock from another beat instance,
+	// we'll need to read from the pid file
+	isLocked, err := lock.fileLock.TryRLock()
+	if err != nil {
+		return fmt.Errorf("unable to lock data path: %w", err)
+	}
+
+	// case: lock could not be obtained.
+	if !isLocked {
+		// if we're here, things are probably unrecoverable, as we've previously checked for a lockfile. Exit.
+		return ErrAlreadyLocked
+	}
+
+	return nil
+}
+
+// Unlock attempts to release the lock on a data path previously acquired via Lock().
+func (lock *Locker) Unlock() error {
+	err := lock.fileLock.Unlock()
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path: %w", err)
+	}
+
+	err = os.Remove(lock.fileLock.Path())
+	if err != nil {
+		return fmt.Errorf("unable to unlock data path: %w", err)
+	}
+	return nil
+}
+
+// ******* private helpers
+
+// createPidfile creates the underling pidfile used by the lock
+// This assumes that no previous lockfile should exist.
+// returns a bool indicating if the file was created, and an error
+func (lock *Locker) createPidfile(pid int) (bool, error) {
+	_, err := os.Stat(lock.filePath)
+	//file already exists, go into repair/handle fail mode
+	if !os.IsNotExist(err) {
+		return false, nil
+	}
+
+	new := pidfile{Pid: pid, WriteTime: time.Now()}
+	encoded, err := json.Marshal(&new)
+	if err != nil {
+		return false, fmt.Errorf("error encoding json for pidfile: %w", err)
+	}
+
+	err = os.WriteFile(lock.filePath, encoded, os.FileMode(0600))
+	if err != nil {
+		return false, fmt.Errorf("error writing pidfile to %s: %w", lock.filePath, err)
+	}
+	return true, nil
+}
+
+// handleFailedLock will attempt to recover from a failed lock operation in a pid-aware way.
+// The point of this is to deal with instances where an improper beat shutdown left us with
+// a pre-existing pidfile for a beat process that no longer exists.
+// The one argument tells the method if a pre-existing lockfile was already found.
+func (lock *Locker) handleFailedLock() error {
+	// read in whatever existing lockfile caused us to fail
+	pf, err := lock.readExistingPidfile()
+	if err != nil {
+		return fmt.Errorf("error reading existing lockfile: %w", err)
+	}
+
+	// Case: the lockfile is locked, but by us. Probably a coding error,
+	// and probably hard to do
+	if pf.Pid == os.Getpid() {
+		return fmt.Errorf("lockfile for beat has been locked twice by the same PID, potential bug.")
+	}
+
+	// Check to see if the PID found in the pidfile exists.
+	// Checking if the pid exists in a cross-platform way is more fraught than it looks.
+	// we have internal metrics libraries we can use for this, but all those will use APIs
+	// dedicated to gathering extended process info and metrics, which can come with extra permissions hurdles,
+	// making those methods more likely to return an error.
+	exists, err := process.PidExistsWithContext(context.Background(), int32(pf.Pid))
```

Comment:
Pid doesn't have to be 32bit.  On linux:

```
 /proc/sys/kernel/pid_max (since Linux 2.5.34)
              This  file specifies the value at which PIDs wrap around (i.e., the value in this file is one greater than the maximum PID).  PIDs greater
              than this value are not allocated; thus, the value in this file also acts as a system-wide limit on the  total  number  of  processes  and
              threads.   The default value for this file, 32768, results in the same range of PIDs as on earlier kernels.  On 32-bit platforms, 32768 is
              the maximum value for pid_max.  On 64-bit systems, pid_max can be set to any value up to 2^22 (PID_MAX_LIMIT, approximately 4 million).
              ```

Reply:
Yah, unfortunately that library wants a 32-bit int, but I think I'm gonna need some kind of wrapper here anyway, since CI is choking on the cgo.